File: python/4.py
# A palindromic number reads the same both ways. The largest palindrome 
# made from the product of two 2-digit numbers is 9009 = 91 x 99.
#
# Find the largest palindrome made from the product of two 3-digit numbers.
#
# https://projecteuler.net/problem=4

import string, sys

import commonutils

def main():
    largest = 0

    for i in range(100, 999):
        for j in range(100, 999):
            prod = i * j

            if commonutils.isPalindrome(str(prod)) and prod > largest:
                largest = prod

    print (largest)

if __name__ == '__main__':
    main()

File: c/4.c
/* http://projecteuler.net/problem=4 */
#include <string.h>
#include <stdio.h>

int is_palindrome(char *buf)
{
    int i=0, j=strlen(buf)-1;
    while (i<=j) {
        if (buf[i] != buf[j]) {
            return 0;
        }
        i++;
        j--;
    }

    return 1;
}

int main(int argc, char* argv[])
{
        int i=0, j=0, prod = 0, max=0;
        char str[100];

        for (i=101 ; i<1000 ; i++) {
            for (j=101 ; j<1000 ; j++) {
                prod = i*j;
                sprintf(str, "%d\0", prod);
                if (is_palindrome(str) && (prod>max)) {
                    max = prod;
                }
            }
        }

        printf("max palindrome = %d\n", max);


        return 0;
}
